activerecord/test/cases/store_test.rb

```diff
@@ -191,4 +191,42 @@ class StoreTest < ActiveRecord::TestCase
     second_dump = YAML.dump(loaded)
     assert_equal @john, YAML.load(second_dump)
   end
+
+  test "setting type casted values on build" do
+    jamie = Admin::User.new(name: "Jamie", active: "true", salary: 3.1999, birthday: '2000-01-01')
+    assert_equal true, jamie.active
+    assert_equal 3, jamie.salary
+    assert_equal Date.new(2000,1,1), jamie.birthday
+  end
+
+  test "setting type casted values from db" do
+    jamie = Admin::User.new(name: "Jamie")
+    jamie.active = "t"
+    jamie.salary = 3.1999 
+    jamie.birthday = '01/01/2000'
+    jamie.save!
+
+    jamie = Admin::User.where(name: "Jamie").first
+   
+    assert_equal '2000-01-01', jamie.json_data['birthday']    
```

Comment:
In that case, we just need to swap the implementation around a bit. We should decorate the type of the store itself so for all of the methods on the type, we cast each of the keys. Something like:

```
class TypedStore < DelegateClass(Type::Value)
  def initialize(subtype, accessor_types)
    @accessor_types = accessor_types
    super(subtype)
  end

  def cast(value)
    hash = super
    if hash
      accessor_types.each do |key, type|
        hash[key] = type.cast(hash[key])
      end
    end
    hash
  end

  protected

  attr_reader :accessor_types
end

decorate_attribute_type(accessor, :typed_accessors) do |subtype|
  TypedStore.new(subtype, accessor_types)
end
```

With additional methods for `serialize` and `deserialize` with the same type of delegation. We will probably need to go through the subtypes accessor as well, but I think you can see the direction I'm going